File: test/t_basic/SaveTest.js
/*
 Copyright (c) 2012, Oracle and/or its affiliates. All rights
 reserved.
 
 This program is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public License
 as published by the Free Software Foundation; version 2 of
 the License.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 02110-1301  USA
 */

// TODO save with some but not all fields included

/***** Save no update***/
var t1 = new harness.ConcurrentTest("testSaveNoUpdate");
t1.run = function() {
  var testCase = this;
  // create the domain object 4000
  var object = new global.t_basic(4000, 'Employee 4000', 4000, 4000);
  fail_openSession(testCase, function(session) {
    // key and testCase are passed to fail_verify_t_basic as extra parameters
    session.save(object, function(err, session2) {
      if (err) {
        testCase.fail(err);
        return;
      }
      session2.find(global.t_basic, 4000, fail_verify_t_basic, 4000, testCase, true);
    }, session);
  });
};

/***** Batch save no update ***/
var t2 = new harness.ConcurrentTest("testBatchSaveNoUpdate");
t2.run = function() {
  var testCase = this;
  var start_number = 4010;
  var number_of_objects = 10;
  var i, object;
  
  fail_openSession(testCase, function(session) {
    var batch = session.createBatch();
    // create objects
    for (i = start_number; i < start_number + number_of_objects; ++i) {
      object = new global.t_basic(i, 'Employee ' + i, i, i);
      batch.save(object, function(err) {
        if (err) {
          testCase.appendErrorMessage(err);
        }
      });
    }
    batch.execute(function(err, session2) {
      if (err) {
        testCase.appendErrorMessage(err);
      }
      var batch2 = session2.createBatch();
      for (i = start_number; i < start_number + number_of_objects; ++i) {
        batch2.find(global.t_basic, i, verify_t_basic, i, testCase, true);
      }
      batch2.execute(function(err) {
        if (err) {
          testCase.appendErrorMessage(err);
        }
        testCase.failOnError();
      });
    }, session);
  });
};

/***** Save update***/
var t3 = new harness.ConcurrentTest("testSaveUpdate");
t3.run = function() {
  var testCase = this;
  // load the domain object 4020
  var object = new global.t_basic(4020, 'Employee 4020', 4020, 4020);
  var object2;
  fail_openSession(testCase, function(session) {
    // save object 4020
    session.save(object, function(err, session2) {
      if (err) {
        testCase.fail(err);
        return;
      }
      // now save an object with the same primary key but different magic
      object2 = new global.t_basic(4020, 'Employee 4020', 4020, 4029);
      session2.save(object2, function(err, session3) {
        if (err) {
          testCase.fail(err);
          return;
        }
        session3.find('t_basic', 4020, function(err, object3) {
          // verify that object3 has updated magic field from object2
          testCase.errorIfNotEqual('testSaveUpdate mismatch on magic', 4029, object3.magic);
          testCase.failOnError();
        });
      }, session2);
    }, session);
  });
};

/***** Batch save update ***/
var t4 = new harness.ConcurrentTest("testBatchSaveUpdate");
t4.run = function() {
  var testCase = this;
  var start_number = 4030;
  var number_of_objects = 10;
  // create objects
  
  fail_openSession(testCase, function(session) {
    var batch = session.createBatch();
    for (i = start_number; i < start_number + number_of_objects; i += 2) {
      batch.save(new global.t_basic(i, 'Employee ' + (i + 100), i + 100, i + 100), function(err) {
        if (err) {
          testCase.appendError(err);
        }
      });
    }
    batch.execute(function(err, session2) {
      if (err) {
        testCase.appendErrorMessage(err);
      }
      var batch2 = session2.createBatch();
      // save objects (on top of original objects)
      for (i = start_number; i < start_number + number_of_objects; ++i) {
        batch2.save(new global.t_basic(i, 'Employee ' + i, i, i), function(err) {
          if (err) {
            testCase.appendErrorMessage(err);
          }
        });
      }
      batch2.execute(function(err, session3) {
        if (err) {
         testCase.appendErrorMessage(err); 
        }
        var batch3 = session3.createBatch();
        // find all objects and verify they have the proper state
        for (i = start_number; i < start_number + number_of_objects; ++i) {
          batch3.find(global.t_basic, i, global.verify_t_basic, i, testCase, true);
        }
        batch3.execute(function(err) {
          if (err) {
            testCase.appendErrorMessage(err);
          }
          testCase.failOnError();
        });
      }, session2);
    }, session);
  });
};

/***** Save illegal parameter ***/
var t5 = new harness.ConcurrentTest("testSaveNoUpdate");
t5.run = function() {
  var testCase = this;
  fail_openSession(testCase, function(session) {
    session.save(1, function(err, session2) {
      if (err) {
        testCase.pass();
        return;
      }
      testCase.fail(new Error('SaveTest.t5 illegal argument should fail.'));
    });
  });
};

/***** Save illegal parameter ***/
var t6 = new harness.ConcurrentTest("testSaveNoUpdate");
t6.run = function() {
  var testCase = this;
  fail_openSession(testCase, function(session) {
    session.save('t_basic', function(err, session2) {
      if (err) {
        testCase.pass();
        return;
      }
      testCase.fail(new Error('SaveTest.t6 illegal argument should fail.'));
    });
  });
};

module.exports.tests = [t1, t2, t3, t4, t5, t6];

File: storage/ndb/clusterj/clusterj-test/src/main/java/testsuite/clusterj/SaveTest.java
/*
   Copyright 2010 Sun Microsystems, Inc.
   Use is subject to license terms.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License, version 2.0,
   as published by the Free Software Foundation.

   This program is also distributed with certain software (including
   but not limited to OpenSSL) that is licensed under separate terms,
   as designated in a particular file or component or in included license
   documentation.  The authors of MySQL hereby grant you an additional
   permission to link the program and your derivative works with the
   separately licensed software that they have included with MySQL.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License, version 2.0, for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA
*/

package testsuite.clusterj;

import java.util.ArrayList;
import java.util.List;
import testsuite.clusterj.model.Employee;

public class SaveTest extends AbstractClusterJModelTest {

    private static final int NUMBER_TO_INSERT = 4;
    
    @Override
    public void localSetUp() {
        createSessionFactory();
        session = sessionFactory.getSession();
        createEmployeeInstances(NUMBER_TO_INSERT);
        tx = session.currentTransaction();
        tx.begin();
        session.deletePersistentAll(Employee.class);
        tx.commit();
        tx.begin();
        List<Employee> insertedEmployees = new ArrayList<Employee>();
        for (int i = 0; i < NUMBER_TO_INSERT; ++i) {
            if (i%2 == 0) {
                // only make even employees persistent now
                insertedEmployees.add(employees.get(i));
            }
        }
        session.makePersistentAll(insertedEmployees);
        tx.commit();
        addTearDownClasses(Employee.class);
    }

    public void testSave() {
        tx.begin();
        for (int i = 0; i < NUMBER_TO_INSERT; ++i) {
            Employee e = session.find(Employee.class, i);
            if (e != null) {
                if (i%2 != 0) {
                    error("Employee " + i + " should not exist.");
                }
                // if exists, change age
                e.setAge(NUMBER_TO_INSERT - i);
            } else {
                // if not exist, insert with new age
                if (i%2 == 0) {
                    error("Employee " + i + " should exist.");
                } else {
                    e = employees.get(i);
                    e.setAge(NUMBER_TO_INSERT - i);
                }
            }
            // send the change to the database
            session.savePersistent(e);
        }
        tx.commit();

        // now verify that the changes were committed
        tx.begin();
        for (int i = 0; i < NUMBER_TO_INSERT; ++i) {
            Employee e = session.find(Employee.class, i);
            if (e == null) {
                error("Failed save: employee " + i + " does not exist.");
            } else {
                // verify age
                int expected = NUMBER_TO_INSERT - i;
                int actual = e.getAge();
                if (expected != actual) {
                    error("Failed save: for employee " + i
                            + " expected age " + expected
                            + " actual age " + actual);
                }
            }
        }
        tx.commit();
        failOnError();
    }

    public void testSaveAll() {
        tx.begin();
        List<Employee> emps = new ArrayList<Employee>();
        List<Employee> expectedEmployees = new ArrayList<Employee>();
        for (int i = 0; i < NUMBER_TO_INSERT; ++i) {
            Employee e = session.find(Employee.class, i);
            if (e != null) {
                if (i%2 != 0) {
                    error("Employee " + i + " should not exist.");
                }
                // if exists, change age
                e.setAge(NUMBER_TO_INSERT - i);
            } else {
                // if not exist, insert with new age
                if (i%2 == 0) {
                    error("Employee " + i + " should exist.");
                } else {
                    e = employees.get(i);
                    e.setAge(NUMBER_TO_INSERT - i);
                }
            }
            emps.add(e);
            expectedEmployees.add(e);
        }
        // send the changes to the database
        List<Employee> savedEmployees = (List<Employee>)session.savePersistentAll(emps);
        if (savedEmployees.size() != NUMBER_TO_INSERT) {
            error("Wrong size for saved employees. Expected: " + NUMBER_TO_INSERT
                    + " actual: " + savedEmployees.size());
        }
        for (int i = 0; i < NUMBER_TO_INSERT; ++i) {
            Employee e = expectedEmployees.get(i);
            // verify saved employees
            Employee saved = savedEmployees.get(i);
            if (saved != e) {
                error ("Failed saveAll: employee " + i + " did not match savedEmployees. "
                        + "Expected: " + e.toString() + " hashcode: " + e.hashCode()
                        + " actual: " + saved.toString() + " hashcode: " + saved.hashCode());
            }
        }
        tx.commit();

        // now verify that the changes were committed
        tx.begin();
        for (int i = 0; i < NUMBER_TO_INSERT; ++i) {
            Employee e = session.find(Employee.class, i);
            if (e == null) {
                error("Failed saveAll: employee " + i + " does not exist.");
            } else {
                // verify age
                int expected = NUMBER_TO_INSERT - i;
                int actual = e.getAge();
                if (expected != actual) {
                    error("Failed saveAll: for employee " + i
                            + " expected age " + expected
                            + " actual age " + actual);
                }
            }
        }
        tx.commit();
        failOnError();
    }
}
